File: swift/HelloIroh/WireFormat.swift
import Foundation

enum WireFormat {
    static let alpn: Data = Data("iroh-helloiroh-dot/0".utf8)

    static let positionFrameSize: UInt32 = 8   // x(4) + y(4)

    static func encodePosition(x: Float, y: Float) -> Data {
        var out = Data(count: 8)
        out.withUnsafeMutableBytes { raw in
            let base = raw.baseAddress!
            base.assumingMemoryBound(to: UInt32.self).pointee = x.bitPattern.littleEndian
            base.advanced(by: 4).assumingMemoryBound(to: UInt32.self).pointee = y.bitPattern.littleEndian
        }
        return out
    }

    struct Position {
        let x: Float
        let y: Float
    }

    static func decodePosition(_ data: Data) -> Position? {
        guard data.count == 8 else { return nil }
        return data.withUnsafeBytes { raw in
            let base = raw.baseAddress!
            let xBits = UInt32(littleEndian: base.assumingMemoryBound(to: UInt32.self).pointee)
            let yBits = UInt32(littleEndian: base.advanced(by: 4).assumingMemoryBound(to: UInt32.self).pointee)
            return Position(x: Float(bitPattern: xBits), y: Float(bitPattern: yBits))
        }
    }
}

File: kotlin-android/app/src/main/java/computer/iroh/dot/net/WireFormat.kt
package computer.iroh.dot.net

import java.nio.ByteBuffer
import java.nio.ByteOrder

object WireFormat {
    val ALPN: ByteArray = "iroh-helloiroh-dot/0".toByteArray(Charsets.UTF_8)

    const val POSITION_FRAME_SIZE: UInt = 8u // x(4) + y(4)

    data class Position(val x: Float, val y: Float)

    fun encodePosition(x: Float, y: Float): ByteArray {
        val buf = ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN)
        buf.putFloat(x)
        buf.putFloat(y)
        return buf.array()
    }

    fun decodePosition(body: ByteArray): Position? {
        if (body.size != 8) return null
        val buf = ByteBuffer.wrap(body).order(ByteOrder.LITTLE_ENDIAN)
        return Position(x = buf.float, y = buf.float)
    }
}
